Organize Sample Code (#2912)
nasa-gcn/gcn.nasa.gov: app/routes/docs.sample._index.mdx

---
handle:
  breadcrumb: Sample Code
---

# Sample code for GCN Notices

GCN utilizes [Apache Kafka](https://kafka.apache.org), an open-source distributed event streaming platform, to stream Notices and Circulars in real time.

[GCN Notices](https://gcn.nasa.gov/docs/notices) provide real-time machine readable alerts and are distributed in different formats:

- New GCN Notices are provided in JSON format and are streamed via Kafka. Note that SVOM notices are only available via VOEvent over Kafka and email.
- GCN Classic Notices are provided in text, binary and VOEvent XML, and are distributed both via Kafka and email.

This sample code demonstrates how to handle GCN Notices in both JSON and VOEvent XML formats, and some samples from the FAQs section of the [gcn-kafka-python](https://github.com/nasa-gcn/gcn-kafka-python) repository.

## Working with different GCN Notice Formats:

> - [Kafka Message Consumption](sample/kafka) demonstrates how to consume and manage GCN messages using Kafka clients.

Reviewer: ```suggestion
- [Consuming Kafka Messages](sample/kafka) demonstrates how to consume and manage GCN messages using Kafka clients.
```

---
handle:
  breadcrumb: Sample Code
---

# Sample Code

GCN utilizes [Apache Kafka](https://kafka.apache.org), an open-source distributed event streaming platform, to stream Notices and Circulars in real time.

[GCN Notices](/docs/notices) provide real-time machine readable alerts and are distributed in different formats:

- New GCN Notices are provided in JSON format and are streamed via Kafka, with the exception of SVOM notices which are only available via VOEvent over Kafka and email.
- GCN Classic Notices are provided in text, binary, and VOEvent XML formats, and are distributed both via Kafka and email.

This sample code demonstrates how to handle GCN Notices in both JSON and VOEvent XML formats, and some samples from the FAQs section of the [gcn-kafka-python](https://github.com/nasa-gcn/gcn-kafka-python) repository.

## Working with different GCN Notice Formats

> - [Consuming Kafka Messages](sample/kafka) demonstrates how to consume and manage GCN messages using Kafka clients.

Reviewer 2: ```suggestion
- [Kafka Messages](sample/kafka) demonstrates how to consume and manage GCN messages using Kafka clients.
```

> - [Text Messages](sample/text) demonstrates how to parse a GCN message.

Reviewer 2: ```suggestion
- [Plain Text Notices](sample/text) demonstrates how to parse emailed plain text format Notices.
```

> - [Plain Text Notices](sample/text) demonstrates how to parse emailed plain text format GCN messages.

Reviewer: ```suggestion
- [Plain Text Notices](sample/text) demonstrates how to parse plain text format GCN messages.
```

> - [Plain Text Notices & Circulars](sample/text) demonstrates how to parse emailed plain text format GCN messages.

Author: Emailed plain text format is applicable for Circulars as well.

> - [Plain Text Notices & Circulars](sample/text) demonstrates how to parse emailed plain text format GCN messages.

Reviewer: If they are, then it's a coincidence. Please don't mention Circulars here. This is irrelevant.

Please don't conflate the GCN Classic text format with the email delivery method.

> - [XML Notices](sample/xml) provides guidance for interpreting and extracting data from XML formatted GCN Notices.

Reviewer 2: ```suggestion
- [VOEvent XML Notices](sample/xml) provides guidance for interpreting and extracting data from VOEvent XML formatted GCN Notices.
```

> - [JSON Notices](sample/json) presents sample code for processing GCN Notices encoded in JSON format.

Author: ```suggestion
- [JSON Notices](sample/json) presents sample code for processing GCN Notices encoded in JSON format, and instructions for Producers on how to encode FITS files in JSON format Notices.
```

> - [HEALPix Sky Map](sample/healpix) explains how to interpret and utilize HEALPix format sky maps included in GCN data.

Reviewer 2: ```suggestion
- [HEALPix Sky Map](sample/healpix) explains how to interpret and utilize HEALPix format sky maps included in GCN Notices, and instructions for Producers on how to encode FITS files in JSON format Notices.
```

> - [HEALPix Sky Map](sample/healpix) explains how to interpret and utilize HEALPix format sky maps included in GCN Notices, and instructions for Producers on how to encode FITS files in JSON format Notices.

Reviewer 2: ```suggestion
- [HEALPix Sky Map](sample/healpix) explains how to interpret and utilize HEALPix format sky maps included in GCN Notices.
```

---
handle:
  breadcrumb: Sample Code
---

# Sample code for GCN Notices

This sample code demonstrates how to handle GCN Notices in both JSON and VOEvent XML formats, and some samples from the FAQs section of the [gcn-kafka-python](https://github.com/nasa-gcn/gcn-kafka-python) repository.

## Apache Kafka and GCN Streaming

[Apache Kafka](https://kafka.apache.org) is an open-source distributed event streaming platform used by GCN to stream Notices and Circulars in real time.

## GCN Notices Formats and Distributions

[GCN Notices](https://gcn.nasa.gov/docs/notices) provide real-time machine readable alerts and are distributed in different formats:

- New GCN Notices are provided in JSON format and are streamed via Kafka. Note that SVOM notices are only available via VOEvent over Kafka and email.
- GCN Classic Notices are provided in text, binary and VOEvent XML, and are distributed both via Kafka and email.

> 

Reviewer 2: Remove headers for "Apache Kafka ..." and "GCN Notices Formats ...". Then move "This sample code ..." to after them, so that this is all background info for the sample code.

> ## The following sections provide guidance on handling different GCN Notices formats:

Reviewer 3: ```suggestion
## Additional Information
The following sections provide guidance on handling different GCN Notices formats:
```

> - [Kafka Usage](sample/kafka) explains Kafka's functionality

Reviewer 3: ```suggestion
- [Kafka Consumption](sample/kafka) explains Kafka's functionality and how to consume notices.
```

Author: I have modified, please check latest.

> - [XML Data](sample/xml) guides users for working with XML Notices

Reviewer 3: ```suggestion
- [XML Data](sample/xml) guides users on working with XML Notices.
```

> - [JSON Data](sample/json) provides sample code for handling JSON-based Notices

Reviewer 3: ```suggestion
- [JSON Data](sample/json) provides sample code for handling JSON-based Notices.
```

> - [HEALPix](sample/healpix) explains how to work with HEALPix sky maps

Reviewer 3: ```suggestion
- [HEALPix](sample/healpix) explains how to work with HEALPix sky maps.
```

---
handle:
  breadcrumb: Sample Code
---

# Sample Code

Here is a collection of functions and example code that may be useful
when working with GCN Notices. This sample code demonstrates how to handle GCN Notices in both JSON and VOEvent XML formats, and some samples from the FAQs section of the [gcn-kafka-python](https://github.com/nasa-gcn/gcn-kafka-python) repository.

## Apache Kafka and GCN Streaming

[Apache Kafka](https://kafka.apache.org) is an open-source distributed event streaming platform used by GCN to stream Notices and Circulars in real time.

## GCN Notices Formats and Distributions

GCN Notices provide real-time machine readable alerts and are distributed in different formats:

- New GCN Notices are provided in JSON format and are streamed via Kafka. Note that SVOM notices are only available via VOEvent over Kafka and email.

> - GCN Classic Notices are provided in text, binary and VOEvent XML, and are distributed both via Kafka and email.

Reviewer 3: Perhaps we could clarify the distribution methods available for the two types on the main notices docs page and then link to it here?

---
handle:
  breadcrumb: Sample Code
---

> # Apache Kafka and GCN Streaming

Reviewer 2: This shouldn't be the level 1 header.  The main header should "Sample Code".

[Apache Kafka](https://kafka.apache.org) is an open-source distributed event streaming platform used by GCN to stream Notices and Circulars in real time.

## GCN Notices Formats and Distributions

GCN Notices provide real-time machine readable alerts and are distributed in different formats:

> - New GCN Notices are provided in JSON format and are streamed via Kafka. Note SVOM available via VOEvent over Kafka.

Reviewer 2: ```suggestion
- New GCN Notices are provided in JSON format and are streamed via Kafka. Note that SVOM notices are only available via VOEvent over Kafka and email.
```

Reviewer 2: [Adding email for SVOM](https://github.com/nasa-gcn/gcn-email/pull/281)

Author: > [Adding email for SVOM](https://github.com/nasa-gcn/gcn-email/pull/281)

Shall we wait on this?
I am committing suggestions.

> - GCN Classic Notices are formatted in text, binary and VOEvent XML, are distributed both via Kafka and email.

Reviewer 2: ```suggestion
- GCN Classic Notices are provided in text, binary and VOEvent XML, and are distributed both via Kafka and email.
```

## Sample Code

Here is a collection of functions and example code that may be useful
when working with GCN Notices. This sample code demonstrates how to handle GCN Notices in both JSON and VOEvent XML formats, and some samples from the FAQs section of the [gcn-kafka-python](https://github.com/nasa-gcn/gcn-kafka-python) repository.

### The following sections provide guidance on handling different GCN Notices formats:

- [Kafka Usage](sample/kafka) explains Kafka's functionality
- [XML Data](sample/xml) guides users for working with XML Notices
- [JSON Data](sample/json) provides sample code for handling JSON-based Notices
- [HEALPix](sample/healpix) explains how to work with HEALPix sky maps

> To contribute your own ideas, make a GitHub pull request to add it to [the Markdown source for this document](https://github.com/nasa-gcn/gcn.nasa.gov/blob/CodeSamples/app/routes/docs.sample._index.mdx), or [contact us](/contact).

Reviewer 2: ```suggestion
To contribute your own sample code, make a GitHub pull request to add it to [the Markdown source for this document](https://github.com/nasa-gcn/gcn.nasa.gov/blob/CodeSamples/app/routes/docs.sample._index.mdx), or [contact us](/contact).
```

---
handle:
  breadcrumb: Sample Code
---

# Apache Kafka and GCN Streaming

> [Apache Kafka](https://kafka.apache.org) is an open-source distributed event streaming platform used by thousands of companies for high-performance data pipelines, streaming analytics, data integration, and mission-critical applications.

Reviewer 2: ```suggestion
[Apache Kafka](https://kafka.apache.org) is an open-source distributed event streaming platform used by GCN to stream Notices and Circulars in real time.
```

> NASA leverages Kafka extensively, with notable implementations including the General Coordinates Network (GCN) at the Goddard Space Flight Center, the Complex Event Processor for Deep Space Network at Jet Propulsion Laboratory, and other federal agencies. These organizations use self-managed Kafka brokers, either Apache Kafka or the Confluent Platform, to support their data-intensive operations. GCN is sponsoring [FedRAMP](https://www.fedramp.gov) authorization for [Confluent Cloud](https://www.confluent.io/confluent-cloud/) to make it easy for NASA and other federal agencies to deploy Kafka software-as-a-service.

Reviewer 2: ```suggestion
```

> NASA leverages Kafka extensively, with notable implementations including the General Coordinates Network (GCN) at the Goddard Space Flight Center, the Complex Event Processor for Deep Space Network at Jet Propulsion Laboratory, and other federal agencies. These organizations use self-managed Kafka brokers, either Apache Kafka or the Confluent Platform, to support their data-intensive operations. GCN is sponsoring [FedRAMP](https://www.fedramp.gov) authorization for [Confluent Cloud](https://www.confluent.io/confluent-cloud/) to make it easy for NASA and other federal agencies to deploy Kafka software-as-a-service.
> 
> GCN’s Kafka cluster is simple yet flexible, it operates on a standard three-broker Kafka cluster, without any custom authentication extensions or vendor lock-in. It uses Confluent Platform but can easily switch to open-source Kafka or managed solutions like Confluent Cloud. With [OpenID Connect (OIDC)](https://openid.net/developers/how-connect-works/) for single sign-on, utilizing [Amazon Cognito](https://aws.amazon.com/cognito/) for authentication, GCN ensures consistency across its web services and Kafka broker, making the system adaptable and secure.

Reviewer 2: ```suggestion
```

## GCN Notices Formats and Distributions

GCN Notices provide real-time machine readable alerts and are distributed in different formats:

> - New GCN Notices are provided in JSON format and are streamed via Kafka.

Reviewer 2: ```suggestion
- New GCN Notices are provided in JSON format (GCN Unified Schema) and are streamed via Kafka.
```

> - Classic GCN Notices are formatted in XML as well, are distributed both via Kafka and email.

Reviewer 2: ```suggestion
- GCN Classic Notices are formatted in text, binary and VOEvent XML, are distributed both via Kafka and email.
```

## Sample Code

Here is a collection of functions and example code that may be useful

> when working with GCN Notices. The sample code here demonstrates how to handle GCN Notices in both JSON and XML formats, and some samples from the FAQs section of the [gcn-kafka-python](https://github.com/nasa-gcn/gcn-kafka-python) repository.

Reviewer 2: ```suggestion
when working with GCN Notices. This sample code demonstrates how to handle GCN Notices in both JSON and VOEvent XML formats, and some samples from the FAQs section of the [gcn-kafka-python](https://github.com/nasa-gcn/gcn-kafka-python) repository.
```

> ### The following sections provide guidance on handling different GCN Notice formats:

Author: ```suggestion
### The following sections provide guidance on handling different GCN Notices formats:
```
